pkg/remotewrite/remotewrite.go

```diff
@@ -94,71 +93,183 @@ func (o *Output) flush() {
-	promTimeSeries := make([]prompb.TimeSeries, 0)
+func (o *Output) convertToPbSeries(samplesContainers []metrics.SampleContainer) []prompb.TimeSeries {
+	// The seen map is required because the samples containers
+	// could have several samples for the same time series
+	// in this way we can aggregate and flush them in a unique value
+	// without overloading the remote write endpoint.
+	seen := make(map[string]struct{})
 
 	for _, samplesContainer := range samplesContainers {
 		samples := samplesContainer.GetSamples()
 
 		for _, sample := range samples {
-			// Prometheus remote write treats each label array in TimeSeries as the same
-			// for all Samples in those TimeSeries (https://github.com/prometheus/prometheus/blob/03d084f8629477907cab39fc3d314b375eeac010/storage/remote/write_handler.go#L75).
-			// But K6 metrics can have different tags per each Sample so in order not to
-			// lose info in tags or assign tags wrongly, let's store each Sample in a different TimeSeries, for now.
-			// This approach also allows to avoid hard to replicate issues with duplicate timestamps.
-
-			labels, err := tagsToLabels(sample.Tags, o.config)
-			if err != nil {
-				o.logger.Error(err)
+			timeSeriesKey := timeSeriesKey(sample.Metric, sample.Tags)
+			swm, ok := o.tsdb[timeSeriesKey]
+			if !ok {
+				swm = &seriesWithMeasure{
+					TimeSeries: TimeSeries{
+						Metric: sample.Metric,
+						Tags:   sample.Tags,
+					},
+					Measure: sinkByType(sample.Metric.Type),
+					Latest:  sample.Time,
+				}
+				o.tsdb[timeSeriesKey] = swm
 			}
-
-			if newts, err := o.metrics.transform(o.mapping, sample, labels); err != nil {
-				o.logger.Error(err)
-			} else {
-				promTimeSeries = append(promTimeSeries, newts...)
+			swm.Measure.Add(sample)
+			if sample.Time.After(swm.Latest) {
+				swm.Latest = sample.Time
 			}
+			seen[timeSeriesKey] = struct{}{}
 		}
+	}
+
+	pbseries := make([]prompb.TimeSeries, 0, len(seen))
+	now := time.Now()
+	for s := range seen {
+		pbseries = append(pbseries, o.tsdb[s].MapPrompb(now)...)
+	}
+	return pbseries
+}
+
+type seriesWithMeasure struct {
+	Measure metrics.Sink
+	Latest  time.Time
 
-		// Do not blow up if remote endpoint is overloaded and responds too slowly.
-		// TODO: consider other approaches
-		if flushTooLong && len(promTimeSeries) > 150000 {
-			break
+	// TImeSeries will be replaced with the native k6 version
+	// when it will be available.
+	TimeSeries
+
+	// TODO: maybe add some caching for the mapping?
+}
+
+func (swm seriesWithMeasure) MapPrompb(t time.Time) []prompb.TimeSeries {
+	var newts []prompb.TimeSeries
+
+	mapMonoSeries := func(s TimeSeries, t time.Time) prompb.TimeSeries {
+		return prompb.TimeSeries{
+			Labels: append(MapTagSet(swm.Tags), prompb.Label{
+				Name:  "__name__",
+				Value: fmt.Sprintf("%s%s", defaultMetricPrefix, swm.Metric.Name),
+			}),
+			Samples: []prompb.Sample{
+				{Timestamp: timestamp.FromTime(t)},
```

Comment:
```suggestion
			Samples: []prompb.Sample{
			        // here timestamp is rounded down from ns to ms precision
				{Timestamp: timestamp.FromTime(t)},
```
I meant something like this. Sorry for not being more clear yesterday :sweat_smile:

Reply:
This comment plus more tests have triggered good thoughts on my side :+1:  and I caught some issues so I added a better optimization where we use directly the truncated time otherwise some cases wouldn't be covered as described in the comments.
Now the truncate operation is explicit in the logic so we shouldn't need one more comment.'

Reply:
Glad to hear that!

:+1: I'll re-read the PR on Mon and let you know if I notice anything else.